### DISCUSSION TITLE
queue count of a single destination

### DISCUSSION THREAD

**Discussion by JoeFox82** [nextgenhealthcare/connect#5837] [General]
The class ChannelUtil has a function to get the count of all queued messages in the channel. But is there a way to see in which destination the messages are queued or to get the queue count for a single destination?

**Comment by jonbartels** [comment]
@JoeFox82 can you show the code you are currently calling please?

What is your use case? Are you making message processing decisions in code or are you doing general reporting?

In broad strokes, yes there is. If you can see it in the dashboard or messages view, you can get it in code.

The approaches are:

1. HTTP calls to the Mirth API - I think you want one of the stats endpoints
2. Local calls to the same Controllers that the API calls anyway - Once you know the HTTP endpoint you can peek at the code and see what controllers it calls and what objects it gets
3. Use various util methods - You found this already
4. Query the DB - I'm a fan of [this approach](https://gist.github.com/jonbartels/b961574b2043b628f1b0fd96f440179b) but you have to understand the DB structure. This is most useful for broad reporting.
5. Various commercial offerings for stats - Zen and Innovar both have a dashboard tool.

**Reply by JoeFox82** [reply]
I'd like to send an email, when the queue count is more then 100 for a destination. Actually I'm using some code that was shared in a mirth forum. Here is a part of the code, that is stored in an JavaScript Reader Channel that is looking for the queue count of a channel

var channel_expression = "SELECT C.NAME, C.ID, D.LOCAL_CHANNEL_ID FROM D_CHANNELS D, CHANNEL C WHERE D.CHANNEL_ID = C.ID;";
var channel_result = dbConn.executeCachedQuery(channel_expression);

while (channel_result.next())
{
  	var ChannelID   		= channel_result.getString('ID');

	if(!ChannelUtil.isChannelDeployed(ChannelID))
		continue;

	if(ChannelUtil.getQueuedCount(ChannelID) > maxQueuedMessages)
	{
		// send issue message via mail
	}
}

**Reply by jonbartels** [reply]
That code should work but it is getting the total queued count for a channel, not for a specific destination.

For this kind of logic why not just use the alerts functionality?

**Reply by JoeFox82** [reply]
Yes, it works for the total count. But I want to have the queue count of a destination and I don't know how to do this :-(

**Reply by jonbartels** [reply]
Take a look at https://github.com/nextgenhealthcare/connect/blob/e0db5497aabb94c41071e4a098580e660a1d5d8e/server/src/com/mirth/connect/server/controllers/DefaultChannelController.java#L740 

`com.mirth.connect.donkey.server.controllers.ChannelController.getInstance().getConnectorMessageCount(channelId, serverId, metaDataId, status);` will return the data you want. 

- You already have the channelId.
- serverId is nullable (I dug through the code and null is fine here)
- metaDataId is the number of the destination
- status should be `com.mirth.connect.donkey.model.message.Status.QUEUED`

**Comment by tonygermano** [comment]
http://javadocs.mirthcorp.com/connect/4.0.0/user-api/com/mirth/connect/server/userutil/ChannelUtil.html

`ChannelUtil.getQueuedCount()` has two versions. One takes only a channelId, and the other takes both a channelId and a connector metaDataId.

**Comment by JoeFox82** [comment]
I solved the problem using this code :-)

```
	var controller = com.mirth.connect.server.controllers.ControllerFactory.getFactory().createChannelController();
	for each (destinationConnector in controller.getDeployedChannelById(ChannelID).getDestinationConnectors().toArray())
	{			
		// Now we will check to see if the the stats.queued is over 100
		//   If so we need to compare the channel Name to get some values to use for the alerts
		if(ChannelUtil.getQueuedCount(ChannelID, destinationConnector.getMetaDataId()) > maxQueuedMessages)
		{
                      // do something
                }
       }
```
